apps/aecore/lib/aecore/util/state_trees.ex

```diff
@@ -0,0 +1,95 @@
+defmodule Aecore.Util.StateTrees do
+  alias Aecore.Account.AccountStateTree
+  alias Aecore.Contract.CallStateTree
+  alias Aecore.Contract.ContractStateTree
+  alias Aecore.Channel.ChannelStateTree
+  alias Aecore.Naming.NamingStateTree
+
+  @moduledoc """
+  Abstract module defining functions for known Tree types
+  """
+  defmacro __using__(_) do
+    quote location: :keep do
+      alias Aecore.Contract.{Contract, ContractStateTree}
+      alias Aecore.Chain.Identifier
+      alias Aecore.Naming.{Name, NameCommitment}
+      alias Aecore.Util.StateTrees
+      alias Aeutil.PatriciaMerkleTree
+      alias Aeutil.Serialization
+      alias MerklePatriciaTree.Trie
+
+      @typedoc "Hash of the tree"
+      @type hash :: binary()
+
+      @spec init_empty :: Trie.t()
+      def init_empty do
+        PatriciaMerkleTree.new(StateTrees.tree_type(__MODULE__))
+      end
+
+      @spec put(Trie.t(), binary(), map()) :: Trie.t()
+      def put(tree, key, value) do
+        serialized_state = Serialization.rlp_encode(value)
+        PatriciaMerkleTree.enter(tree, key, serialized_state)
+      end
+
+      @spec get(Trie.t(), binary()) :: :none | map()
+      def get(tree, key) do
+        case PatriciaMerkleTree.lookup(tree, key) do
+          {:ok, serialized_value} ->
+            {:ok, deserialized_value} = Serialization.rlp_decode_anything(serialized_value)
+
+            case deserialized_value do
```

Comment:
This `case ... do` prevents `Aecore.Util.StateTrees` from being abstract. You can extract it to a function and each ***StateTree module will implement modification of corresponding structure.